File: apps/studio/src/app/shared/models/device.ts
import { NormalizedLayout } from '../utils/layout/models';

export interface Capabilities {
  volatileApply: boolean;
  commit: boolean;
  layouts: boolean;
  keymap: boolean;
  sequences: boolean;
  actionTypes?: string[];
  targetCounts?: {
    keys: number;
    encoders: number;
    controls: number;
  };
  encoderDescriptors?: Array<{
    id: number;
    detent?: boolean;
    velocity?: boolean;
    analog?: boolean;
    ticksPerRev?: number | null;
    minRate?: number | null;
    maxRate?: number | null;
  }>;
}

export interface DeviceInfo {
  id: string;
  name: string;
  transport: string;
  vendorId?: string;
  productId?: string;
  manufacturer?: string | null;
  product?: string | null;
  serial?: string | null;
  firmwareVersion?: string;
  fingerprint?: string | null;
  interfaces?: DeviceInterface[];
  capabilityLevel?: string | null;
  definitionLinked?: boolean | null;
  definitionFingerprint?: string | null;
}

export interface DeviceInterface {
  usagePage: number;
  usage: number;
  interfaceNumber?: number | null;
  path?: string | null;
  label?: string | null;
}

export interface ViaProbe {
  viaDetected: boolean;
  viaProtocolVersion?: number | null;
  writeLen?: number | null;
  readLen?: number | null;
  timeoutMs?: number | null;
  firstBytes?: string | null;
}

export interface ViaState {
  viaDetected: boolean;
  protocolVersion?: number | null;
  layers: number;
  layerCount?: number;
  keymap: string[][][];
  encoderMap?: string[][][];
  macros?: string[];
}

export interface Step {
  id: number;
  name: string;
  op: string;
  arg?: string;
  class?: number;
}

export type Binding =
  | { type: 'none' }
  | { type: 'sequenceRef'; sequenceId: string; meta?: Record<string, unknown> }
  | { type: 'simpleAction'; action: string; arg?: string; meta?: Record<string, unknown> }
  | { type: 'inlineSequence'; steps: Step[]; meta?: Record<string, unknown> }
  | { type: 'program'; path: string; meta?: Record<string, unknown> };

export interface BindingEntry {
  targetId: string;
  binding: Binding;
  layerId?: number;
}

export interface Sequence {
  id: string;
  profileId: string;
  name: string;
  steps: Step[];
  meta?: Record<string, unknown>;
}

export interface LayerState {
  id: number;
  bindings: BindingEntry[];
}

export interface DeviceState {
  profileId: string;
  layers: LayerState[];
  revision?: number;
  checksum?: number;
}

export interface Profile {
  id: string;
  name: string;
  layers: LayerState[];
}

export interface ProfileBundle {
  sessionId: string;
  definitionFingerprint?: string | null;
  device: DeviceInfo;
  capabilities: Capabilities;
  profile: Profile;
  layout: NormalizedLayout | null;
  targets: string[];
  sequences: Sequence[];
  committedState?: DeviceState | null;
  appliedState?: DeviceState | null;
  stagedState?: DeviceState | null;
  bindings?: BindingEntry[];
}

export interface StatusSnapshot {
  running: string | null;
  ramLoaded: boolean;
  dirty: boolean;
  committedState?: DeviceState | null;
  appliedState?: DeviceState | null;
  stagedState?: DeviceState | null;
  definitionFingerprint?: string | null;
}

File: apps/studio/src-tauri/src/models/device.rs
use serde::{Deserialize, Serialize};

use super::binding::BindingEntry;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceInfo {
  pub id: String,
  pub name: String,
  #[serde(default)]
  pub transport: String,
  #[serde(rename = "vendorId", default)]
  pub vendor_id: Option<String>,
  #[serde(rename = "productId", default)]
  pub product_id: Option<String>,
  #[serde(rename = "firmwareVersion", default)]
  pub firmware_version: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Capabilities {
  #[serde(rename = "volatileApply")]
  pub volatile_apply: bool,
  pub commit: bool,
  pub layouts: bool,
  pub keymap: bool,
  pub scripts: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LayerState {
  pub id: i32,
  pub bindings: Vec<BindingEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceState {
  #[serde(rename = "profileId")]
  pub profile_id: String,
  pub layers: Vec<LayerState>,
  #[serde(default)]
  pub revision: Option<i32>,
  #[serde(default)]
  pub checksum: Option<u32>,
}
